File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {
	static int N;
	static int M;
	static int ans;
	static char[][] grid;
	static boolean[][][] visited;
	static KNode start;
	static int[] dx = {-1, 1, 0, 0};
	static int[] dy = {0, 0, -1, 1};
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		String[] temp = br.readLine().split(" ");
		
		N = Integer.parseInt(temp[0]);
		M = Integer.parseInt(temp[1]);
		
		grid = new char[N][M];
		visited = new boolean[64][N][M];
		
		for(int i = 0; i < N; i++) {
			String str = br.readLine();
			for(int j = 0; j < M; j++) {
				grid[i][j] = str.charAt(j);
				if(grid[i][j] == '0') {
					start = new KNode(i, j, 0, 0);
				}
			}
		}
		
		System.out.println(bfs());
	}
	
	public static int bfs() {
		Queue<KNode> q = new LinkedList<>();
		q.add(new KNode(start.x, start.y, 0, 0));
		visited[0][start.x][start.y] = true;
		
		while(!q.isEmpty()) {
			KNode node = q.poll();
			int x = node.x;
			int y = node.y;
			int dist = node.dist;
			int key = node.key;
			
			if(grid[x][y] == '1') {
				return dist;
			}
			
			for(int i = 0; i < 4; i++) {
				int nx = x + dx[i];
				int ny = y + dy[i];
				
				if(isValid(nx, ny) && grid[nx][ny] != '#' && !visited[key][nx][ny]) {
					if(grid[nx][ny] == '.' || grid[nx][ny] == '0' || grid[nx][ny] == '1') {
						visited[key][nx][ny] = true;
						q.offer(new KNode(nx, ny, dist + 1, key));
					}
					else if(grid[nx][ny] >= 'a' && grid[nx][ny] <= 'f') {
						int newkey = 1 << (grid[nx][ny] - 'a');
						newkey = newkey | key;
						if(!visited[newkey][nx][ny]) {
							visited[key][nx][ny] = true;
							visited[newkey][nx][ny] = true;
							q.offer(new KNode(nx, ny, dist + 1, newkey));
						}
					}
					else if(grid[nx][ny] >= 'A' && grid[nx][ny] <= 'F') {
						int door = 1 << (grid[nx][ny] - 'A');
						if((key & door) > 0) {
							visited[key][nx][ny] = true;
							q.offer(new KNode(nx, ny, dist + 1, key));
						}
					}
				}
			}
		}
		return -1;
	}
	
	public static boolean isValid(int nx, int ny) {
		if(nx < 0 || nx >= N || ny < 0 || ny >= M) {
			return false;
		}
		return true;
	}
}

class KNode{
	int x;
	int y;
	int dist;
	int key;
	
	public KNode(int x, int y, int dist, int key) {
		this.x = x;
		this.y = y;
		this.dist = dist;
		this.key = key;
	}
}

File: 백준/Gold/1194. 달이 차오른다， 가자．/달이 차오른다， 가자．.py
from collections import deque

class Node:
    def __init__(self, x, y, dist, key):
        self.x = x
        self.y = y
        self.dist = dist
        self.key = key
N, M = map(int, input().split())

graph = []

for i in range(N):
    graph.append(list(input()))
    if '0' in graph[i]:
        start = Node(i, graph[i].index('0'), 0, 0)
        graph[i][graph[i].index('0')] = '.'

visited = [[[False] * M for _ in range(N)] for _ in range(64)]
alphaL = ['a', 'b', 'c', 'd', 'e', 'f']
alphaU = ['A', 'B', 'C', 'D', 'E', 'F']

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(start):
    q = deque()
    q.append(start)

    while q:
        node = q.popleft()
        x = node.x
        y = node.y
        dist = node.dist
        key = node.key

        if graph[x][y] == '1':
            print(dist)
            exit(0)

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < N and 0 <= ny < M and graph[nx][ny] != '#' and not visited[key][nx][ny]:
                if graph[nx][ny] in alphaL:
                    newkey = key | (1 << (ord(graph[nx][ny]) - ord('a')))
                    if not visited[newkey][nx][ny]:
                        visited[newkey][nx][ny] = True
                        visited[key][nx][ny] = True
                        q.append(Node(nx, ny, dist + 1, newkey))
                elif graph[nx][ny] in alphaU:
                    door = 1 << (ord(graph[nx][ny]) - ord('A'))

                    if key & door > 0:
                        if not visited[key][nx][ny]:
                            visited[key][nx][ny] = True
                            q.append(Node(nx, ny, dist + 1, key))
                else:
                    q.append(Node(nx, ny, dist + 1, key))
                    visited[key][nx][ny] = True

bfs(start)

print(-1)
